Plugins/BuildServerIntegration/VstsAndTfsIntegration/TfsApiHelper.cs

```diff
@@ -37,81 +37,81 @@ private HttpClient InitializeHttpClient(string serverUrl, string projectName, st
-                        if (buildDefinitionNameFilter.IsMatch(buildDefinition.Name))
-                        {
-                            buildDefs.Add(buildDefinition);
-                        }
-                    }
-                }
+                return JsonConvert.DeserializeObject<T>(json);
+            }
+        }
+
+        private async Task<string> GetBuildDefinitionsAsync(string buildDefinitionNameFilter)
+        {
+            var isNotFiltered = string.IsNullOrWhiteSpace(buildDefinitionNameFilter);
+            var buildDefinitionUriFilter = isNotFiltered ? string.Empty : "&name=" + buildDefinitionNameFilter;
+            var buildDefinitions = await HttpGetAsync<ListWrapper<BuildDefinition>>(BuildDefinitionsUrl + buildDefinitionUriFilter);
+            if (buildDefinitions.Count != 0)
+            {
+                return GetBuildDefinitionsIds(buildDefinitions.Value);
+            }
+
+            if (isNotFiltered)
+            {
+                return null;
+            }
+
+            buildDefinitions = await HttpGetAsync<ListWrapper<BuildDefinition>>(BuildDefinitionsUrl);
+
+            var tfsBuildDefinitionNameFilter = new Regex(buildDefinitionNameFilter, RegexOptions.Compiled);
+            return GetBuildDefinitionsIds(buildDefinitions.Value.Where(b => tfsBuildDefinitionNameFilter.IsMatch(b.Name)));
+        }
+
+        private string GetBuildDefinitionsIds(IEnumerable<BuildDefinition> buildDefinitions)
+        {
+            if (buildDefinitions != null && buildDefinitions.Any())
+            {
+                return string.Join(",", buildDefinitions.Select(b => b.Id));
             }
 
-            return buildDefs;
+            return null;
         }
 
-        public void ConnectToTfsServer(string serverUrl, string teamCollection, string projectName, string restApiToken, Regex buildDefinitionNameFilter = null)
+        public void ConnectToTfsServer(string serverUrl, string teamCollection, string projectName, string restApiToken, string buildDefinitionNameFilter)
         {
             try
             {
                 _httpClient = InitializeHttpClient(serverUrl, projectName, restApiToken);
 
-                var taskServerInit = ThreadHelper.JoinableTaskFactory.RunAsync(async () => await GetBuildDefinitionsAsync(teamCollection, projectName, buildDefinitionNameFilter));
-                _buildDefinitions = taskServerInit.Join();
+                var taskServerInit = ThreadHelper.JoinableTaskFactory.RunAsync(async () => await GetBuildDefinitionsAsync(buildDefinitionNameFilter));
+                _buildDefinitionsToQuery = taskServerInit.Join();
             }
             catch (Exception ex)
             {
                 Trace.WriteLine(ex.Message);
             }
         }
 
-        public async Task<IReadOnlyList<Build>> QueryBuildsAsync(DateTime? sinceDate, bool? running)
+        public async Task<IEnumerable<Build>> QueryBuildsAsync(DateTime? sinceDate, bool? running)
```

Comment:
this method must be wrapped in `try..catch` otherwise it has a chance to bring down the app